README.md
# `typed-arena-any-vec`

[![](https://docs.rs/typed-arena/badge.svg)](https://docs.rs/typed-arena-any-vec/)
[![](https://img.shields.io/crates/v/typed-arena-any-vec.svg)](https://crates.io/crates/typed-arena-any-vec)
[![](https://img.shields.io/crates/d/typed-arena-any-vec.svg)](https://crates.io/crates/typed-arena-any-vec)

**[typed-arena](https://docs.rs/typed-arena)** but supports any type of `Vec` backing, including [`SmallVec`](https://docs.rs/smallvec), [`ArrayVec`](https://docs.rs/arrayvec), and [`SliceVec`](https://docs.rs/slicevec). This means that you can create areans on data which you are mutably borrowing or data which is not stored on the heap.

Do note that if you're putting arenas on the stack, make sure that they're small, as the stack doesn't have much memory.

## `GrowVec`

The backing type is a `GrowVec` provided in this module. This vector only needs to support insertion, as allocated objects are only destroyed all at once when the vec itself is dropped.

`GrowVec` is implemented for these external crates. To use, add the crate's name as a feature:

- [`smallvec`](https://docs.rs/smallvec)
- [`arrayvec`](https://docs.rs/arrayvec)
- [`slicevec`](https://docs.rs/slicevec)

If you have another external trait you want `GrowVec` to support, create a PR.

## Examples (from typed-arena)

```rust
use typed_arena_any_vec::Arena;
use arrayvec::ArrayVec;

struct Monster {
    level: u32,
}

fn fun() {
    let monsters = Arena::new(ArrayVec::<Monster, 4>::new());

    let goku = monsters.alloc(Monster { level: 9001 });
    assert!(goku.level > 9000).unwrap();
}
```

### Safe Cycles

All allocated objects get the same lifetime, so you can safely create cycles
between them. This can be useful for certain data structures, such as graphs
and trees with parent pointers.

```rust
use std::cell::Cell;
use typed_arena_any_vec::Arena;
use slicevec::SliceVec;

struct CycleParticipant<'a> {
    other: Cell<Option<&'a CycleParticipant<'a>>>,
}

struct CapacityError;

fn fun(backing: &mut [CycleParticipant<'_>]) -> Result<(), CapacityError> {
    let arena = Arena::new(SliceVec::new(backing));

    let a = arena.alloc(CycleParticipant { other: Cell::new(None) }).map_err(|_| CapacityError);
    let b = arena.alloc(CycleParticipant { other: Cell::new(None) }).map_err(|_| CapacityError);

    a.other.set(Some(b));
    b.other.set(Some(a));
}
```


CHANGELOG.md
## Unreleased

Released YYYY/MM/DD.

### Added

* TODO (or remove section if none)

### Changed

* TODO (or remove section if none)

### Deprecated

* TODO (or remove section if none)

### Removed

* TODO (or remove section if none)

### Fixed

* TODO (or remove section if none)

### Security

* TODO (or remove section if none)

--------------------------------------------------------------------------------

## 0.1.0

Released 2022/08/04.

src/test.rs
use super::*;
#[cfg(any(feature = "arrayvec", feature = "slicevec"))]
use std::cell::Cell;
#[cfg(any(feature = "slicevec"))]
use std::mem::MaybeUninit;
#[cfg(feature = "arrayvec")]
use arrayvec::ArrayVec;
#[cfg(feature = "slicevec")]
use slicevec::SliceVec;

#[derive(Debug, Clone)]
struct DropTracker<'a>(&'a Cell<u32>);

impl<'a> Drop for DropTracker<'a> {
    fn drop(&mut self) {
        self.0.set(self.0.get() + 1);
    }
}

#[derive(Debug, Clone)]
struct Node<'a, 'b: 'a>(Option<&'a Node<'a, 'b>>, u32, DropTracker<'b>);

#[cfg(feature = "arrayvec")]
#[test]
fn array_arena() {
    let drop_counter = Cell::new(0);
    {
        let arena = Arena::new(ArrayVec::<_, 2>::new());

        let mut node = arena.alloc(Node(None, 1, DropTracker(&drop_counter))).unwrap();
        node = arena.alloc(Node(Some(node), 2, DropTracker(&drop_counter))).unwrap();

        assert_eq!(node.1, 2);
        assert_eq!(node.0.unwrap().1, 1);
        assert!(node.0.unwrap().0.is_none());
        assert_eq!(arena.len(), 2);

        let error = arena.alloc(Node(Some(node), 3, DropTracker(&drop_counter))).unwrap_err();
        let error_elem = error.element();
        assert_eq!(error_elem.1, 3);

        assert_eq!(drop_counter.get(), 0);
        drop(error_elem);
        assert_eq!(drop_counter.get(), 1);

        drop(node);
    }
    assert_eq!(drop_counter.get(), 3);
    drop_counter.set(0);

    {
        let arena = Arena::new(ArrayVec::<_, 25>::new());

        let mut node = arena.alloc(Node(None, 1, DropTracker(&drop_counter))).unwrap();
        node = arena.alloc(Node(Some(node), 2, DropTracker(&drop_counter))).unwrap();
        node = arena.alloc(Node(Some(node), 3, DropTracker(&drop_counter))).unwrap();

        assert_eq!(node.1, 3);
        assert_eq!(node.0.unwrap().1, 2);
        assert_eq!(arena.len(), 3);

        let mut node = arena.alloc(Node(None, 4, DropTracker(&drop_counter))).unwrap();
        node = arena.alloc(Node(Some(node), 5, DropTracker(&drop_counter))).unwrap();

        assert_eq!(drop_counter.get(), 0);
        assert_eq!(node.1, 5);
        assert_eq!(node.0.unwrap().1, 4);
        assert!(node.0.unwrap().0.unwrap().0.is_none());
    }
    assert_eq!(drop_counter.get(), 7);
}

#[cfg(feature = "slicevec")]
#[test]
fn slice_arena() {
    let drop_counter_buffer2 = Cell::new(0);
    let mut buffer2 = MaybeUninit::uninit_array::<25>();
    for elem in buffer2.iter_mut() {
        elem.write(Node(None, 100, DropTracker(&drop_counter_buffer2)));
    }
    // let mut buffer2 = unsafe { MaybeUninit::array_assume_init(buffer2) };

    let drop_counter = Cell::new(0);
    {
        let mut buffer1 = MaybeUninit::uninit_array::<2>();
        for elem in buffer1.iter_mut() {
            elem.write(Node(None, 10, DropTracker(&drop_counter)));
        }
        let mut buffer1 = unsafe { MaybeUninit::array_assume_init(buffer1) };        let arena = Arena::new(SliceVec::new(&mut buffer1));

        let mut node = arena.alloc(Node(None, 1, DropTracker(&drop_counter))).unwrap();
        node = arena.alloc(Node(Some(node), 2, DropTracker(&drop_counter))).unwrap();

        assert_eq!(node.1, 2);
        assert_eq!(node.0.unwrap().1, 1);
        assert!(node.0.unwrap().0.is_none());
        assert_eq!(arena.len(), 2);

        let error = arena.alloc(Node(Some(node), 3, DropTracker(&drop_counter))).unwrap_err();
        assert_eq!(error.1, 3);

        assert_eq!(drop_counter.get(), 2);
    }

    assert_eq!(drop_counter.get(), 5);
    drop_counter.set(0);

    /* {
        let arena = Arena::new(SliceVec::new(&mut buffer2));

        let node1 = arena.alloc(Node(None, 1, DropTracker(&drop_counter))).unwrap();
        let node2 = arena.alloc(Node(Some(node1), 2, DropTracker(&drop_counter))).unwrap();
        let node3 = arena.alloc(Node(Some(node2), 3, DropTracker(&drop_counter))).unwrap();

        assert_eq!(node3.1, 3);
        assert_eq!(node3.0.unwrap().1, 2);
        assert_eq!(node2.1, 2);
        assert_eq!(arena.len(), 3);

        let node4 = arena.alloc(Node(None, 4, DropTracker(&drop_counter))).unwrap();
        let node5 = arena.alloc(Node(Some(node4), 5, DropTracker(&drop_counter))).unwrap();

        assert_eq!(drop_counter.get(), 0);
        assert_eq!(node5.1, 5);
        assert_eq!(node5.0.unwrap().1, 4);
        assert!(node5.0.unwrap().0.unwrap().0.is_none());

        assert_eq!(drop_counter_buffer2.get(), 7);
    }
    assert_eq!(drop_counter.get(), 0);
    assert_eq!(drop_counter_buffer2.get(), 7);

    drop(buffer2);

    assert_eq!(drop_counter.get(), 7);
    assert_eq!(drop_counter_buffer2.get(), 100); */
}

#[test]
#[cfg(feature = "stable_deref_trait")]
fn ensure_into_vec_maintains_order_of_allocation() {
    let arena = Arena::new(Vec::new());
    for &s in &["t", "e", "s", "t"] {
        arena.alloc(String::from(s)).unwrap();
    }
    let vec = arena.into_vec();
    assert_eq!(vec, vec!["t", "e", "s", "t"]);
}

#[test]
#[cfg(feature = "arrayvec")]
fn test_is_send() {
    fn assert_is_send<T: Send>(_: T) {}

    // If `T` is `Send`, ...
    assert_is_send(42_u32);

    // Then `Arena<T>` is also `Send`.
    let arena: Arena<u32, ArrayVec<u32, 5>> = Arena::new(ArrayVec::new());
    assert_is_send(arena);
}

#[derive(Debug, PartialEq, Eq)]
struct NonCopy(usize);

#[test]
#[cfg(feature = "arrayvec")]
fn iter_mut_full_capacity() {
    const MAX: usize = 1000;

    let mut arena = Arena::new(ArrayVec::<_, 1000>::new());
    for i in 0..MAX {
        arena.alloc(NonCopy(i)).unwrap();
    }

    let mut iter = arena.iter_mut();
    assert_eq!(iter.len(), MAX);

    for i in 0..MAX {
        assert_eq!(Some(&mut NonCopy(i)), iter.next());
    }

    assert!(iter.is_empty());
    assert_eq!(None, iter.next());
}

#[test]
#[cfg(feature = "arrayvec")]
fn iter_mut_not_full_capacity() {
    const MAX: usize = 1000;

    let mut arena = Arena::new(ArrayVec::<_, 2000>::new());
    for i in 0..MAX {
        arena.alloc(NonCopy(i)).unwrap();
    }

    let mut iter = arena.iter_mut();
    assert_eq!(iter.len(), MAX);

    for i in 0..MAX {
        assert_eq!(Some(&mut NonCopy(i)), iter.next());
    }

    assert!(iter.is_empty());
    assert_eq!(None, iter.next());
}

src/grow_vec.rs
#[cfg(feature = "frozenvec")]
use core::alloc::Allocator;
#[cfg(feature = "frozenvec")]
use core::convert::Infallible;
#[cfg(feature = "frozenvec")]
use core::marker::PhantomData;
#[cfg(feature = "arrayvec")]
use core::mem::{MaybeUninit};
#[cfg(any(feature = "arrayvec", feature = "slicevec"))]
use transmute::transmute;
#[cfg(any(feature = "arrayvec", feature = "slicevec"))]
use core::ptr::{addr_of, addr_of_mut};
#[cfg(feature = "frozenvec")]
use stable_deref_trait::StableDeref;
#[cfg(feature = "arrayvec")]
use arrayvec::ArrayVec;
#[cfg(feature = "slicevec")]
use slicevec::SliceVec;

/// A vector which supports mutable indexing and insertion, and you can insert into the vector
/// while indexed values have live references without UB.
///
/// Does not need to support removal.
pub trait GrowVec<T> {
    /// The type which actually supports insertion with active references.
    type Raw;
    /// Error when the vector is full. Use [Infallible] if the vector can grow to the system memory.
    type CapacityError;

    /// Convert from the type which actually supports insertion with active references.
    fn from_raw(raw: Self::Raw) -> Self;

    /// Convert to the type which actually supports insertion with active references.
    fn into_raw(self) -> Self::Raw;

    /// Gets the number of pushed values from a mutable pointer to `Self`.
    ///
    /// SAFETY: the pointer must point to an initialized instance.
    unsafe fn len_from_ptr(this: *const Self::Raw) -> usize;

    /// Gets the mutable pointer of a value at an index a mutable pointer.
    ///
    /// SAFETY: the pointer must point to an initialized instance.
    unsafe fn index_mut_from_ptr(this: *mut Self::Raw, idx: usize) -> *mut T;

    /// Tries to push a value from a mutable pointer to `Self`.
    /// Returns [CapacityError] if the vector is full.
    ///
    /// SAFETY: the pointer must point to an initialized instance.
    unsafe fn push_from_ptr(this: *mut Self::Raw, value: T) -> Result<(), Self::CapacityError>;
}

#[cfg(feature = "frozenvec")]
impl<T: StableDeref, A: Allocator> GrowVec<T> for Vec<T, A> {
    type Raw = Vec<T, A>;
    type CapacityError = Infallible;

    fn from_raw(raw: Self::Raw) -> Self {
        raw
    }

    fn into_raw(self) -> Self::Raw {
        self
    }

    unsafe fn len_from_ptr(this: *const Self::Raw) -> usize {
        // SAFETY: We can freely dereference the entire vec because all values returned from index_mut_from_ptr
        // should be converted into (*mut T::Target)s, so when they get converted into references,
        // they should point to data outside of the vec.
        // Therefore, &mut *this and &mut *this.index_mut_from_ptr(idx) can exist simultaneously.
        let this = &*this;
        this.len()
    }

    unsafe fn index_mut_from_ptr(this: *mut Self::Raw, idx: usize) -> *mut T {
        let this = &mut *this;
        &mut this[idx] as *mut T
    }

    unsafe fn push_from_ptr(this: *mut Self::Raw, value: T) -> Result<(), Self::CapacityError> {
        let this = &mut *this;
        this.push(value);
        Ok(())
    }
}

#[cfg(feature = "arrayvec")]
#[doc(hidden)]
pub struct _ArrayVec<T, const CAP: usize> {
    // the `len` first elements of the array are initialized
    xs: [MaybeUninit<T>; CAP],
    len: u32,
}

#[cfg(feature = "arrayvec")]
impl<T, const CAP: usize> GrowVec<T> for ArrayVec<T, CAP> {
    type Raw = _ArrayVec<T, CAP>;
    type CapacityError = arrayvec::CapacityError<T>;

    fn from_raw(raw: Self::Raw) -> Self {
        // SAFETY: have the same struct definition.
        // Technically this is actually unsafe and UB, and there is no way to access private struct fields.
        // But in practice this is ok
        unsafe { transmute::<_ArrayVec<T, CAP>, ArrayVec<T, CAP>>(raw) }
    }

    fn into_raw(self) -> Self::Raw {
        // SAFETY: have the same struct definition.
        unsafe { transmute::<ArrayVec<T, CAP>, _ArrayVec<T, CAP>>(self) }
    }

    unsafe fn len_from_ptr(this: *const Self::Raw) -> usize {
        addr_of!((*this).len).read() as usize
    }

    unsafe fn index_mut_from_ptr(this: *mut Self::Raw, idx: usize) -> *mut T {
        (addr_of_mut!((*this).xs) as *mut T).add(idx)
    }

    unsafe fn push_from_ptr(this: *mut Self::Raw, value: T) -> Result<(), Self::CapacityError> {
        let len = addr_of!((*this).len).read();
        if len == CAP as u32 {
            Err(arrayvec::CapacityError::new(value))
        } else {
            (addr_of_mut!((*this).xs) as *mut T).add(len as usize).write(value);
            addr_of_mut!((*this).len).write(len + 1);
            Ok(())
        }
    }
}


#[cfg(feature = "slicevec")]
struct _SliceVec<'a, T> {
    storage: &'a mut [T],
    len: usize,
}

#[cfg(feature = "slicevec")]
#[doc(hidden)]
pub struct SliceVecRaw<'a, T> {
    storage: *mut [T],
    len: usize,
    lifetime: PhantomData<&'a ()>
}

#[cfg(feature = "slicevec")]
impl<'a, T> GrowVec<T> for SliceVec<'a, T> {
    type Raw = SliceVecRaw<'a, T>;
    type CapacityError = T;

    fn from_raw(raw: Self::Raw) -> Self {
        // SAFETY: lifetime of raw is stored, just separately
        let raw = _SliceVec {
            storage: unsafe { &mut *raw.storage },
            len: raw.len,
        };
        // SAFETY: have the same struct definition.
        unsafe { transmute::<_SliceVec<T>, SliceVec<T>>(raw) }
    }

    fn into_raw(self) -> Self::Raw {
        // SAFETY: Have the same definition
        let this = unsafe { transmute::<SliceVec<'a, T>, _SliceVec<'a, T>>(self) };
        SliceVecRaw {
            storage: this.storage as *mut _,
            len: this.len,
            lifetime: PhantomData
        }
    }

    unsafe fn len_from_ptr(this: *const Self::Raw) -> usize {
        addr_of!((*this).len).read() as usize
    }

    unsafe fn index_mut_from_ptr(this: *mut Self::Raw, idx: usize) -> *mut T {
        (addr_of!((*this).storage).read() as *mut T).add(idx)
    }

    unsafe fn push_from_ptr(this: *mut Self::Raw, value: T) -> Result<(), Self::CapacityError> {
        let len = addr_of!((*this).len).read();
        let storage = addr_of!((*this).storage).read();
        let (storage, capacity) = storage.to_raw_parts();
        if len == capacity {
            Err(value)
        } else {
            (storage as *mut T).add(len as usize).write(value);
            addr_of_mut!((*this).len).write(len + 1);
            Ok(())
        }
    }
}

src/lib.rs
#![feature(ptr_metadata)]
#![feature(exact_size_is_empty)]
#![cfg_attr(all(test, feature = "slicevec"), feature(maybe_uninit_uninit_array))]
#![cfg_attr(all(test, feature = "slicevec"), feature(maybe_uninit_array_assume_init))]
#![cfg_attr(feature = "frozenvec", feature(allocator_api))]
#![doc = include_str!("../README.md")]

#![deny(missing_docs)]
#![cfg_attr(not(any(feature = "std", test)), no_std)]

#[cfg(any(feature = "std", test))]
extern crate core;
#[cfg(any(feature = "arrayvec", feature = "slicevec"))]
extern crate transmute;
#[cfg(feature = "frozenvec")]
extern crate stable_deref_trait;
#[cfg(feature = "arrayvec")]
extern crate arrayvec;
#[cfg(feature = "slicevec")]
extern crate slicevec;

use core::str;
use core::cell::UnsafeCell;
use core::ptr::slice_from_raw_parts_mut;

mod grow_vec;
#[cfg(test)]
mod test;

pub use grow_vec::*;

/// An arena of objects of type `T`.
///
/// ## Example
///
/// ```
/// use typed_arena_any_vec::Arena;
/// use arrayvec::ArrayVec;
///
/// struct Monster {
///     level: u32,
/// }
///
/// fn fun() {
///     let monsters = Arena::new(ArrayVec::<Monster, 5>::new());
///
///     let vegeta = monsters.alloc(Monster { level: 9001 });
///     assert!(vegeta.level > 9000);
/// }
/// ```
pub struct Arena<T, V: GrowVec<T>> {
    // Must be wrapped in UnsafeCell so Rust lets us have active mutable references.
    // while Arena is also active. We can perform actions on V::Raw without actually dereferencing it.
    backing: UnsafeCell<V::Raw>,
}

impl<T, V: GrowVec<T>> Arena<T, V> {
    /// Construct a new arena.
    ///
    /// ## Example
    ///
    /// ```
    /// use typed_arena_any_vec::Arena;
    /// use arrayvec::ArrayVec;
    ///
    /// fn fun() {
    ///    let arena = Arena::new(ArrayVec::<usize, 5>::new());
    /// #  arena.alloc(1).unwrap();
    /// }
    /// ```
    pub fn new(backing: V) -> Arena<T, V> {
        Arena {
            backing: UnsafeCell::new(backing.into_raw()),
        }
    }

    /// Return the size of the arena
    ///
    /// This is useful for using the size of previous typed arenas to build new typed arenas with large enough spaces.
    ///
    /// ## Example
    ///
    /// ```
    /// use typed_arena_any_vec::Arena;
    /// use arrayvec::ArrayVec;
    ///
    /// let arena = Arena::new(ArrayVec::<_, 5>::new());
    /// let a = arena.alloc(1);
    /// let b = arena.alloc(2);
    ///
    /// assert_eq!(arena.len(), 2);
    /// ```
    pub fn len(&self) -> usize {
        unsafe { V::len_from_ptr(self.backing.get()) }
    }

    /// Allocates a value in the arena, and returns a mutable reference
    /// to that value. Returns a capacity error if the vector is full.
    ///
    /// ## Example
    ///
    /// ```
    /// use typed_arena_any_vec::Arena;
    /// use arrayvec::ArrayVec;
    ///
    /// let arena = Arena::new(ArrayVec::<_, 5>::new());
    /// let x = arena.alloc(42);
    /// assert_eq!(*x, 42);
    /// ```
    #[inline]
    pub fn alloc(&self, value: T) -> Result<&mut T, V::CapacityError> {
        unsafe {
            V::push_from_ptr(self.backing.get(), value)?;
            Ok(&mut *V::index_mut_from_ptr(self.backing.get(), V::len_from_ptr(self.backing.get()) - 1))
        }
    }

    /// Convert this `Arena` into a `V`.
    ///
    /// Items in the resulting `V` appear in the order that they were
    /// allocated in.
    ///
    /// ## Example
    ///
    /// ```
    /// use typed_arena_any_vec::Arena;
    /// use arrayvec::ArrayVec;
    ///
    /// fn fun() {
    ///   let arena = Arena::new(ArrayVec::<_, 5>::new());
    ///
    ///   arena.alloc("a").unwrap();
    ///   arena.alloc("b").unwrap();
    ///   arena.alloc("c").unwrap();
    ///
    ///   let easy_as_123 = arena.into_vec();
    ///
    ///   assert_eq!(easy_as_123, vec!["a", "b", "c"]);
    /// }
    /// ```
    pub fn into_vec(self) -> V {
        V::from_raw(self.backing.into_inner())
    }

    /// Returns an iterator that allows modifying each value.
    ///
    /// Items are yielded in the order that they were allocated.
    ///
    /// ## Example
    ///
    /// ```
    /// use typed_arena_any_vec::Arena;
    /// use arrayvec::ArrayVec;
    ///
    /// #[derive(Debug, PartialEq, Eq)]
    /// struct Point { x: i32, y: i32 };
    ///
    /// fn fun() {
    ///   let mut arena = Arena::new(ArrayVec::<_, 5>::new());
    ///
    ///   arena.alloc(Point { x: 0, y: 0 }).unwrap();
    ///   arena.alloc(Point { x: 1, y: 1 }).unwrap();
    ///
    ///   for point in arena.iter_mut() {
    ///       point.x += 10;
    ///   }
    ///
    ///   let points = arena.into_vec();
    ///
    ///   assert_eq!(points, vec![Point { x: 10, y: 0 }, Point { x: 11, y: 1 }]);
    /// }
    /// ```
    ///
    /// ## Immutable Iteration
    ///
    /// Note that there is no corresponding `iter` method. Access to the arena's contents
    /// requries mutable access to the arena itself.
    ///
    /// ```compile_fail
    /// use typed_arena_any_vec::Arena;
    /// use arrayvec::ArrayVec;
    ///
    /// fn fun() {
    ///   let mut arena = Arena::new(ArrayVec::<_, 5>::new());
    ///   let x = arena.alloc(1).unwrap();
    ///
    ///   // borrow error!
    ///   for i in arena.iter_mut() {
    ///       println!("i: {}", i);
    ///   }
    ///
    ///   // borrow error!
    ///   *x = 2;
    /// }
    /// ```
    #[inline]
    pub fn iter_mut(&mut self) -> IterMut<'_, T, V> {
        // Has to explicitly be before we get_mut on backing
        let len = self.len();
        IterMut {
            idx: 0,
            len,
            ptr: self.backing.get_mut(),
        }
    }
}

impl<V: GrowVec<u8>> Arena<u8, V> {
    /// Allocates a string slice and returns a mutable reference to it.
    ///
    /// This is on `Arena<u8>`, because string slices use byte slices (`[u8]`) as their backing
    /// storage.
    ///
    /// # Example
    ///
    /// ```
    /// use typed_arena_any_vec::Arena;
    /// use arrayvec::ArrayVec;
    ///
    /// fn fun() {
    ///   let arena: Arena<u8, ArrayVec<u8, 25>> = Arena::new(ArrayVec::new());
    ///   let hello = arena.alloc_str("Hello world");
    ///   assert_eq!("Hello world", hello);
    /// }
    /// ```
    #[inline]
    pub fn alloc_str(&self, s: &str) -> Result<&mut str, V::CapacityError> {
        // TODO: optimize if the compiler doesn't
        let start_idx = self.len();
        let bytes = s.bytes();
        let len = bytes.len();
        for byte in bytes {
            self.alloc(byte)?;
        }
        let buffer = unsafe {
            &mut *slice_from_raw_parts_mut(
                V::index_mut_from_ptr(self.backing.get(), start_idx),
                len
            )
        };
        // SAFETY: can't fail because we got from utf8
        Ok(unsafe { str::from_utf8_unchecked_mut(buffer) })
    }
}

impl<T, V: GrowVec<T> + Default> Default for Arena<T, V> {
    fn default() -> Self {
        Self::new(V::default())
    }
}

/// Mutable arena iterator.
///
/// This struct is created by the [`iter_mut`](struct.Arena.html#method.iter_mut) method on [Arenas](struct.Arena.html).
pub struct IterMut<'a, T: 'a, V: GrowVec<T> + 'a> {
    idx: usize,
    len: usize,
    // We could store *mut V::Raw and lifetime separately, because we're only using *mut V::Raw
    ptr: &'a mut V::Raw,
}

impl<'a, T: 'a, V: GrowVec<T> + 'a> Iterator for IterMut<'a, T, V> {
    type Item = &'a mut T;

    fn next(&mut self) -> Option<&'a mut T> {
        if self.idx == self.len {
            None
        } else {
            Some(unsafe { &mut *V::index_mut_from_ptr(self.ptr as *mut V::Raw, self.idx) })
        }
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        let remaining = self.len - self.idx;
        (remaining, Some(remaining))
    }
}

impl<'a, T: 'a, V: GrowVec<T> + 'a> ExactSizeIterator for IterMut<'a, T, V> {
    fn len(&self) -> usize {
        self.len - self.idx
    }

    fn is_empty(&self) -> bool {
        self.len == self.idx
    }
}